File: src/main/java/com/nickcoblentz/montoya/utils/CopyRequestResponseContextMenuProvider.java
package com.nickcoblentz.montoya.utils;

import burp.api.montoya.MontoyaApi;
import burp.api.montoya.core.ToolType;
import burp.api.montoya.http.message.HttpRequestResponse;
import burp.api.montoya.http.message.requests.HttpRequest;
import burp.api.montoya.http.message.responses.HttpResponse;
import burp.api.montoya.ui.contextmenu.*;
import com.nickcoblentz.montoya.LogLevel;
import com.nickcoblentz.montoya.MontoyaLogger;

import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.Clipboard;
import java.awt.datatransfer.ClipboardOwner;
import java.awt.datatransfer.StringSelection;
import java.awt.datatransfer.Transferable;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.*;
import java.util.List;

public class CopyRequestResponseContextMenuProvider implements ContextMenuItemsProvider, ClipboardOwner, ActionListener {
    private final List<Component> _MenuItemList;
    private MontoyaApi _api;
    private ContextMenuEvent _Event;
    private WebSocketContextMenuEvent _WSEvent;
    private int _EventType=0;
    private static int _EventTypeHTTP=1;
    private static int _EventTypeWS=2;
    private static String _CopyRequestAndResponseName = "Full Rq/Rs";
    private static String _CopyRequestAndResponseHeaderName = "Full Rq, Rs Header";
    private static String _CopyURLAndResponseName = "URL, Rs";
    private static String _CopyURLAndResponseHeaderName = "URL, Rs Header";
    private static String _CopyRequestAndResponseIncludeName = "Full Rq/Rs (incl. all)";
    private static String _CopyURLAndResponseHeaderIncludeName = "URL, Rs Header (incl. all)";
    private static String _CopyResponseBodyOnlyName = "Rs Body Only";
    private static List<String> _StripHeadersForTheseCommands = List.of(_CopyRequestAndResponseName,_CopyRequestAndResponseHeaderName,_CopyURLAndResponseName,_CopyURLAndResponseHeaderName);
    private JMenuItem _CopyRequestAndResponseJMenuItem = new JMenuItem(_CopyRequestAndResponseName);
    private JMenuItem _CopyRequestAndResponseHeaderJMenuItem = new JMenuItem(_CopyRequestAndResponseHeaderName);
    private JMenuItem _CopyURLAndResponseJMenuItem = new JMenuItem(_CopyURLAndResponseName);
    private JMenuItem _CopyURLAndResponseHeadersJMenuItem = new JMenuItem(_CopyURLAndResponseHeaderName);
    private JMenuItem _CopyRequestAndResponseIncludeJMenuItem = new JMenuItem(_CopyRequestAndResponseIncludeName);
    private JMenuItem _CopyURLAndResponseHeaderIncludeJMenuItem = new JMenuItem(_CopyURLAndResponseHeaderIncludeName);
    private JMenuItem _CopyResponseBodyOnlyJMenuItem = new JMenuItem(_CopyResponseBodyOnlyName);

    private Set<String> _requestHeadersToStrip = Set.of(
            "Sec-Ch-Ua",
            "Sec-Ch-Ua-Mobile",
            "Sec-Ch-Ua-Full-Version",
            "Sec-Ch-Ua-Arch",
            "Sec-Ch-Ua-Platform",
            "Sec-Ch-Ua-Platform-Version",
            "Sec-Ch-Ua-Model",
            "Sec-Ch-Ua-Bitness",
            "Sec-Ch-Ua-Wow64",
            "Sec-Ch-Ua-Full-Version-List",
            "Upgrade-Insecure-Requests",
            "Sec-Fetch-Site",
            "Sec-Fetch-Mode",
            "Sec-Fetch-User",
            "Sec-Fetch-Dest",
            "Accept-Language",
            "Accept-Encoding",
            "Accept",
            "Priority");

    private Set<String> _responseHeadersToStrip = Set.of(
            "Accept-Ch",
            "P3p",
            "Cache-Control",
            "Pragma",
            "Expires",
            "X-Frame-Options",
            "X-Robots-Tag",
            "X-XSS-Protection",
            "X-Content-Type-Options",
            "Content-Security-Policy",
            "Strict-Transport-Security");

    MontoyaLogger logger = new MontoyaLogger(_api, LogLevel.DEBUG);

    public CopyRequestResponseContextMenuProvider(MontoyaApi api)
    {
        _api=api;
        _CopyRequestAndResponseHeaderJMenuItem.addActionListener(this);
        _CopyRequestAndResponseJMenuItem.addActionListener(this);
        _CopyURLAndResponseJMenuItem.addActionListener(this);
        _CopyURLAndResponseHeadersJMenuItem.addActionListener(this);
        _CopyRequestAndResponseIncludeJMenuItem.addActionListener(this);
        _CopyURLAndResponseHeaderIncludeJMenuItem.addActionListener(this);
        _CopyResponseBodyOnlyJMenuItem.addActionListener(this);
        _MenuItemList = new ArrayList<Component>();
        _MenuItemList.add(_CopyRequestAndResponseJMenuItem);
        _MenuItemList.add(_CopyRequestAndResponseHeaderJMenuItem);
        _MenuItemList.add(_CopyURLAndResponseJMenuItem);
        _MenuItemList.add(_CopyURLAndResponseHeadersJMenuItem);
        _MenuItemList.add(_CopyRequestAndResponseIncludeJMenuItem);
        _MenuItemList.add(_CopyURLAndResponseHeaderIncludeJMenuItem);
        _MenuItemList.add(_CopyResponseBodyOnlyJMenuItem);
    }
    @Override
    public List<Component> provideMenuItems(ContextMenuEvent event) {
        _Event=event;
        _EventType=_EventTypeHTTP;
        if(!event.selectedRequestResponses().isEmpty() || event.messageEditorRequestResponse().isPresent())
        {
            return _MenuItemList;
        }
        return Collections.emptyList();
    }

    @Override
    public List<Component> provideMenuItems(WebSocketContextMenuEvent event) {
        _WSEvent=event;
        _EventType=_EventTypeWS;
        if(!event.selectedWebSocketMessages().isEmpty() || event.messageEditorWebSocket().isPresent())
        {
            return _MenuItemList;
        }
        return Collections.emptyList();
    }

    @Override
    public List<Component> provideMenuItems(AuditIssueContextMenuEvent event) {
        return Collections.emptyList();
    }

    public String surroundWithMarkdown(String s) {
        return String.format("```http\n%s\n```\n\n", s.trim());
    }


    @Override
    public void actionPerformed(ActionEvent e)
    {
        StringBuilder copyMe;
        if(_EventType==_EventTypeWS)
        {
            copyMe = handleWSEvent(e);
        }
        else
        {
            copyMe = handleHTTPEvent(e);
        }

        Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard();
        Transferable transferable = new StringSelection(copyMe.toString());
        clipboard.setContents(transferable, this);

    }

    public StringBuilder handleWSEvent(ActionEvent e)
    {
        StringBuilder copyMe = new StringBuilder();
        List<WebSocketMessage> targetWebSocketMessages;
        if(!_WSEvent.selectedWebSocketMessages().isEmpty())
        {
            targetWebSocketMessages = _WSEvent.selectedWebSocketMessages();
        }
        else if(_WSEvent.messageEditorWebSocket().isPresent())
        {
            targetWebSocketMessages = new LinkedList<WebSocketMessage>();
            targetWebSocketMessages.add(_WSEvent.messageEditorWebSocket().get().webSocketMessage());
        }
        else
        {
            return copyMe;
        }

        if(!targetWebSocketMessages.isEmpty())
        {
            for (WebSocketMessage wsMessage : targetWebSocketMessages)
            {
                if(wsMessage!=null && wsMessage.upgradeRequest()!=null)
                {
                    copyMe.append(String.format("**%s", wsMessage.upgradeRequest().url()));
                    copyMe.append(String.format(" (%s)**\n\n", wsMessage.direction().toString()));
                    copyMe.append(this.surroundWithMarkdown(_api.utilities().byteUtils().convertToString(wsMessage.payload().getBytes())));
                }
            }
        }

        return copyMe;
    }

    public StringBuilder handleHTTPEvent(ActionEvent e)
    {
        StringBuilder copyMe = new StringBuilder();
        List<HttpRequestResponse> targetRequestResponses;
        if(!_Event.selectedRequestResponses().isEmpty())
        {
            targetRequestResponses = _Event.selectedRequestResponses();
        }
        else if (_Event.messageEditorRequestResponse().isPresent())
        {
            targetRequestResponses = new LinkedList<HttpRequestResponse>();
            targetRequestResponses.add(_Event.messageEditorRequestResponse().get().requestResponse());
        }
        else
        {
            return copyMe;
        }

        if (!targetRequestResponses.isEmpty()) {
            for (HttpRequestResponse requestResponse : targetRequestResponses) {

                if (requestResponse.request() != null) {
                    if(!e.getActionCommand().equals(_CopyResponseBodyOnlyName))
                        copyMe.append(String.format("**%s**\n\n", requestResponse.request().url()));
                    if(e.getActionCommand().equals(this._CopyRequestAndResponseHeaderName) || e.getActionCommand().equals(this._CopyRequestAndResponseName) || e.getActionCommand().equals(this._CopyRequestAndResponseIncludeName)) {
                        HttpRequest strippedRequest = requestResponse.request();
                        if(stripHeaders(e.getActionCommand())) {
                            for (String headerName : _requestHeadersToStrip) {
                                while (strippedRequest.hasHeader(headerName)) {
                                    strippedRequest = strippedRequest.withRemovedHeader(headerName);
                                }
                            }
                        }
                        copyMe.append(this.surroundWithMarkdown(_api.utilities().byteUtils().convertToString(strippedRequest.toByteArray().getBytes())));
                    }
                }

                if (requestResponse.response() != null) {

                    HttpResponse strippedResponse = requestResponse.response();
                    if(stripHeaders(e.getActionCommand())) {
                        for (String headerName : _responseHeadersToStrip) {

                            while (strippedResponse.hasHeader(headerName)) {
                                strippedResponse = strippedResponse.withRemovedHeader(headerName);
                            }
                        }
                    }

                    if(e.getActionCommand().equals(_CopyResponseBodyOnlyName)) {

                        copyMe.append(this.surroundWithMarkdown(_api.utilities().byteUtils().convertToString(strippedResponse.body().getBytes())));
                    }
                    else if (e.getActionCommand().equals(this._CopyRequestAndResponseName) || e.getActionCommand().equals(this._CopyURLAndResponseName) || e.getActionCommand().equals(_CopyRequestAndResponseIncludeName)) {
                        copyMe.append(this.surroundWithMarkdown(_api.utilities().byteUtils().convertToString(strippedResponse.toByteArray().getBytes())));
                    } else {
                        byte[] responseHeaders = Arrays.copyOfRange(strippedResponse.toByteArray().getBytes(), 0, strippedResponse.bodyOffset());
                        copyMe.append(this.surroundWithMarkdown(_api.utilities().byteUtils().convertToString(responseHeaders)));
                    }
                }
            }
        }

        return copyMe;
    }

    @Override
    public void lostOwnership(Clipboard clipboard, Transferable contents) {

    }

    private boolean stripHeaders(String command)
    {
        return _StripHeadersForTheseCommands.contains(command);
    }
}

File: src/main/java/com/nickcoblentz/montoya/utils/CopyRequestResponseContextMenuProvider.kt
package com.nickcoblentz.montoya.utils

import burp.api.montoya.MontoyaApi

import burp.api.montoya.http.message.HttpRequestResponse
import burp.api.montoya.ui.contextmenu.*
import com.nickcoblentz.montoya.LogLevel
import com.nickcoblentz.montoya.MontoyaLogger
import java.awt.Component
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.util.*
import javax.swing.JMenuItem

class CopyRequestResponseContextMenuProvider(private val _api: MontoyaApi, private val _copyHandler: CopyRequestResponseHandler) : ContextMenuItemsProvider, ActionListener {
    private val _MenuItemList: List<Component>
    private var currentEvent: ContextMenuEvent? = null
    private var currentWSEvent: WebSocketContextMenuEvent? = null
    private var currentAuditEvent: AuditIssueContextMenuEvent? = null
    private var currentEventType = 0

    private val _EventTypeHTTP = 1
    private val _EventTypeWS = 2
    private val _EventTypeAudit = 3
    private val _CopyRequestAndResponseName = "Full Rq/Rs"
    private val _CopyRequestAndResponseHeaderName = "Full Rq, Rs Header"
    private val _CopyURLAndResponseName = "URL, Rs"
    private val _CopyURLAndResponseHeaderName = "URL, Rs Header"
    private val _CopyRequestAndResponseIncludeName = "Full Rq/Rs (incl. all)"
    private val _CopyURLAndResponseHeaderIncludeName = "URL, Rs Header (incl. all)"
    private val _CopyResponseBodyOnlyName = "Rs Body Only"

    private val _CopyRequestAndResponseJMenuItem = JMenuItem(_CopyRequestAndResponseName)
    private val _CopyRequestAndResponseHeaderJMenuItem = JMenuItem(_CopyRequestAndResponseHeaderName)
    private val _CopyURLAndResponseJMenuItem = JMenuItem(_CopyURLAndResponseName)
    private val _CopyURLAndResponseHeadersJMenuItem = JMenuItem(_CopyURLAndResponseHeaderName)
    private val _CopyRequestAndResponseIncludeJMenuItem = JMenuItem(_CopyRequestAndResponseIncludeName)
    private val _CopyURLAndResponseHeaderIncludeJMenuItem = JMenuItem(_CopyURLAndResponseHeaderIncludeName)
    private val _CopyResponseBodyOnlyJMenuItem = JMenuItem(_CopyResponseBodyOnlyName)


    var logger: MontoyaLogger = MontoyaLogger(_api, LogLevel.DEBUG)

    init {
        _CopyRequestAndResponseHeaderJMenuItem.addActionListener(this)
        _CopyRequestAndResponseJMenuItem.addActionListener(this)
        _CopyURLAndResponseJMenuItem.addActionListener(this)
        _CopyURLAndResponseHeadersJMenuItem.addActionListener(this)
        _CopyRequestAndResponseIncludeJMenuItem.addActionListener(this)
        _CopyURLAndResponseHeaderIncludeJMenuItem.addActionListener(this)
        _CopyResponseBodyOnlyJMenuItem.addActionListener(this)
        _MenuItemList = listOf(_CopyRequestAndResponseJMenuItem,
            _CopyRequestAndResponseHeaderJMenuItem,
            _CopyURLAndResponseJMenuItem,
            _CopyURLAndResponseHeadersJMenuItem,
            _CopyRequestAndResponseIncludeJMenuItem,
            _CopyURLAndResponseHeaderIncludeJMenuItem,
            _CopyResponseBodyOnlyJMenuItem)
    }

    override fun provideMenuItems(event: ContextMenuEvent): List<Component> {
        currentEvent = event
        currentEventType = _EventTypeHTTP
        if (!event.selectedRequestResponses().isEmpty() || event.messageEditorRequestResponse().isPresent) {
            return _MenuItemList
        }
        return emptyList<Component>()
    }

    override fun provideMenuItems(event: WebSocketContextMenuEvent): List<Component> {
        currentWSEvent = event
        currentEventType = _EventTypeWS
        if (!event.selectedWebSocketMessages().isEmpty() || event.messageEditorWebSocket().isPresent) {
            return _MenuItemList
        }
        return emptyList<Component>()
    }

    override fun provideMenuItems(event: AuditIssueContextMenuEvent): List<Component> {
        currentAuditEvent = event
        currentEventType = _EventTypeAudit
        if (event.selectedIssues().isNotEmpty()) {
            return _MenuItemList
        }
        return emptyList<Component>()
    }

    override fun actionPerformed(e: ActionEvent) {
        if (currentEventType == _EventTypeWS) {
            handleWSEvent(e)
        }
        else if(currentEventType == _EventTypeHTTP) {
            handleHTTPEvent(e)
        }
        else {
            handleAuditEvent(e)
        }
    }

    private fun handleWSEvent(e: ActionEvent?) {
        val targetWebSocketMessages: MutableList<WebSocketMessage>
        if (!currentWSEvent!!.selectedWebSocketMessages().isEmpty()) {
            targetWebSocketMessages = currentWSEvent!!.selectedWebSocketMessages()
        } else if (currentWSEvent!!.messageEditorWebSocket().isPresent) {
            targetWebSocketMessages = LinkedList()
            targetWebSocketMessages.add(currentWSEvent!!.messageEditorWebSocket().get().webSocketMessage())
        } else {
            return
        }

        _copyHandler.copyToClipboard(
            _copyHandler.copyItemsWS(targetWebSocketMessages)
        )

    }

    private fun handleAuditEvent(e: ActionEvent?) {

        currentAuditEvent?.let { event ->
            var requestResponses = mutableListOf<HttpRequestResponse>()
            for(selectedIssue in event.selectedIssues()) {
                selectedIssue.requestResponses()?.let {
                    requestResponses.addAll(it)
                }
                /* later
                selectedIssue.collaboratorInteractions()?.let { interactions ->
                    for(interaction in interactions) {
                        interaction.
                    }
                }*/
            }
            _copyHandler.copyToClipboard(
                _copyHandler.copyItemsHTTP(requestResponses)
            )
        }
    }

    private fun handleHTTPEvent(e: ActionEvent) {

        val targetRequestResponses: MutableList<HttpRequestResponse>
        if (!currentEvent!!.selectedRequestResponses().isEmpty()) {
            targetRequestResponses = currentEvent!!.selectedRequestResponses()
        } else if (currentEvent!!.messageEditorRequestResponse().isPresent) {
            targetRequestResponses = LinkedList()
            targetRequestResponses.add(currentEvent!!.messageEditorRequestResponse().get().requestResponse())
        } else {
            return
        }

        _copyHandler.copyToClipboard(
            _copyHandler.copyItemsHTTP(targetRequestResponses,shouldStripHeaders(e.actionCommand),resolveCopyMode(e.actionCommand))
        )
    }

    private fun shouldStripHeaders(actionCommand : String) : Boolean = listOf(_CopyRequestAndResponseName,_CopyRequestAndResponseHeaderName,_CopyURLAndResponseName,_CopyURLAndResponseHeaderName).contains(actionCommand)

    private fun resolveCopyMode(actionCommand : String) : CopyRequestResponseHandler.CopyMode = when(actionCommand) {
        _CopyRequestAndResponseName -> CopyRequestResponseHandler.CopyMode.RequestFullResponseFull
        _CopyRequestAndResponseHeaderName -> CopyRequestResponseHandler.CopyMode.RequestFullResponseHeaders
        _CopyURLAndResponseName -> CopyRequestResponseHandler.CopyMode.URLResponseFull
        _CopyURLAndResponseHeaderName -> CopyRequestResponseHandler.CopyMode.URLResponseHeaders
        _CopyRequestAndResponseIncludeName -> CopyRequestResponseHandler.CopyMode.RequestFullResponseFull
        _CopyURLAndResponseHeaderIncludeName -> CopyRequestResponseHandler.CopyMode.URLResponseHeaders
        _CopyResponseBodyOnlyName -> CopyRequestResponseHandler.CopyMode.ResponseBody
        else -> CopyRequestResponseHandler.CopyMode.RequestFullResponseFull
    }
}